File: Offer-Python/26.二叉搜索树与双向链表.py
## 题目描述
# 输入一棵二叉搜索树，将该二叉搜索树转换成一个排序的双向链表。
# 要求不能创建任何新的结点，只能调整树中结点指针的指向。

# -*- coding:utf-8 -*-
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def __init__(self):
        self.LDR_nodes = []

    def Convert(self, pRootOfTree):
        # write code here
        if not pRootOfTree:
            return None

        self.ConvertLDR(pRootOfTree)

        for i in range(1, len(self.LDR_nodes)-1):
            self.LDR_nodes[i].left = self.LDR_nodes[i-1]
            self.LDR_nodes[i].right = self.LDR_nodes[i+1]
        if len(self.LDR_nodes)>1:
            self.LDR_nodes[0].right = self.LDR_nodes[1]
            self.LDR_nodes[-1].left = self.LDR_nodes[-2]

        return self.LDR_nodes[0]

    def ConvertLDR(self, pRoot):
        if pRoot:
            self.ConvertLDR(pRoot.left)
            self.LDR_nodes.append(pRoot)
            self.ConvertLDR(pRoot.right)


if __name__ == "__main__":
    head = TreeNode(1)
    left = TreeNode(2)
    right = TreeNode(3)
    head.left = left
    head.right = right

    s = Solution()
    s.Convert(head)
    for node in s.LDR_nodes:
        print(node.val)

File: Offer-Cpp/26.二叉搜索树与双向链表.cpp
//
// Created by 陈建虎 on 2019-11-05.
//
//题目描述
//输入一棵二叉搜索树，将该二叉搜索树转换成一个排序的双向链表。要求不能创建任何新的结点，只能调整树中结点指针的指向。

#include <iostream>
#include <vector>

using namespace std;


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;

    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};

class Solution {
public:
    TreeNode *Convert(TreeNode *pRootOfTree) {
        if (pRootOfTree == NULL) {
            return NULL;
        }
        ldr(pRootOfTree);
        for (int i = 0; i < ldrVector.size(); i++) {
            if (i == 0) {
                ldrVector[i]->left = NULL;
            } else {
                ldrVector[i]->left = ldrVector[i - 1];
            }
            if (i == ldrVector.size() - 1) {
                ldrVector[i]->right = NULL;
            } else {
                ldrVector[i]->right = ldrVector[i + 1];
            }
        }
        ldrVector.front()->left = NULL;
        ldrVector.back()->right = NULL;

        return ldrVector.front();
    }

    void ldr(TreeNode *pRootOfTree) {
        if (pRootOfTree != NULL) {
            ldr(pRootOfTree->left);
            ldrVector.push_back(pRootOfTree);
            ldr(pRootOfTree->right);
        }
    }

private:
    vector<TreeNode *> ldrVector;
};